File: 2015/ruby/day18.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end
data = File.read('day18.txt')

sample = <<~TXT
.#.#.#
...##.
#....#
..#...
#.#..#
####..
TXT

def parse(data)
  data.lines.map do |l|
    l.strip.chars.map {|c| c == '.' ? 0 : 1}
  end
end

def visualize(lights)
  lights.map do |row|
    row.map {|l| l == 1 ? '#' : '.'}.join
  end.join("\n")
end

def lit_neighbors(start, x, y)
  cnt = 0
  [x-1, x, x+1].each do |nx|
    next if nx < 0 || nx >= start.size
    [y-1, y, y+1].each do |ny|
      next if ny < 0 || ny >= start[nx].size
      next if nx == x && ny == y

      cnt += start[nx][ny]
    end
  end
  cnt
end

def evaluate(start, corners_on=false)
  rv = []
  start.each_with_index do |row, x|
    rv.push([])
    row.each_with_index do |old_val, y|
      cnt = lit_neighbors(start, x, y)
      new_val = if old_val == 1
        cnt == 2 || cnt == 3 ? 1 : 0
      else
        cnt == 3 ? 1 : 0
      end
      if corners_on && [
        [0,0], [0,row.size - 1], [start.size - 1, 0], [start.size - 1, row.size - 1]
      ].include?([x,y])
        new_val = 1
      end
      rv[-1].push(new_val)
    end
  end
  rv
end

def iterate(start, steps, corners_on=false)
  lights = parse(start)
  steps.times { lights = evaluate(lights, corners_on)}
  lights
end

assert_equal(<<~TXT.strip, visualize(iterate(sample, 1)))
..##..
..##.#
...##.
......
#.....
#.##..
TXT

assert_equal(<<~TXT.strip, visualize(iterate(sample, 2)))
..###.
......
..###.
......
.#....
.#....
TXT

assert_equal(<<~TXT.strip, visualize(iterate(sample, 3)))
...#..
......
...#..
..##..
......
......
TXT

assert_equal(<<~TXT.strip, visualize(iterate(sample, 4)))
......
......
..##..
..##..
......
......
TXT

puts "Part 1 "
puts iterate(data, 100).flatten.sum

puts "Part 2"
puts iterate(data, 100, true).flatten.sum

File: 2015/go/day18/day18.go
package day18

import "reflect"

func Parse(input []string) [][]int {
	rv := make([][]int, 0)
	for _, line := range input {
		v := make([]int, 0)
		for _, c := range []byte(line) {
			if c == '.' {
				v = append(v, 0)
			} else {
				v = append(v, 1)
			}
		}

		rv = append(rv, v)
	}
	return rv
}

func LitNeighbors(start [][]int, x int, y int) int {
	cnt := 0
	for _, nx := range []int{x - 1, x, x + 1} {
		if nx < 0 || nx >= len(start) {
			continue
		}
		for _, ny := range []int{y - 1, y, y + 1} {
			if ny < 0 || ny >= len(start[nx]) {
				continue
			}
			if nx == x && ny == y {
				continue
			}
			cnt += start[nx][ny]
		}
	}
	return cnt
}

func Evaluate(start [][]int, cornersOn bool) [][]int {
	rv := make([][]int, 0)
	for x, row := range start {
		rv = append(rv, make([]int, 0))
		for y, oldVal := range row {
			cnt := LitNeighbors(start, x, y)
			var newVal int
			if oldVal == 1 {
				if cnt == 2 || cnt == 3 {
					newVal = 1
				} else {
					newVal = 0
				}
			} else {
				if cnt == 3 {
					newVal = 1
				} else {
					newVal = 0
				}
			}
			if cornersOn {
				corners := [][]int{
					{0, 0},
					{0, len(row) - 1},
					{len(start) - 1, 0},
					{len(start) - 1, len(row) - 1},
				}
				if Contains(corners, []int{x, y}) {
					newVal = 1
				}
			}
			rv[len(rv)-1] = append(rv[len(rv)-1], newVal)
		}
	}
	return rv
}

func Contains(s [][]int, e []int) bool {
	for _, a := range s {
		if reflect.DeepEqual(a, e) {
			return true
		}
	}
	return false
}

func Iterate(input []string, steps int, cornersOn bool) [][]int {
	lights := Parse(input)
	for i := 0; i < steps; i++ {
		lights = Evaluate(lights, cornersOn)
	}
	return lights
}

func FlatSum(data [][]int) int {
	rv := 0
	for _, r := range data {
		for _, v := range r {
			rv += v
		}
	}
	return rv
}
